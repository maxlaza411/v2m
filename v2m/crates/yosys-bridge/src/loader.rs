use anyhow::{anyhow, bail, ensure, Context, Result};
use serde::Deserialize;
use serde_json::Value;
use std::{collections::BTreeMap, fs, path::Path};

#[derive(Debug, Default, Clone)]
pub struct LoaderOptions {
    pub allow_mem_blackbox: bool,
}

#[derive(Debug)]
pub struct RtlilJson {
    pub top: String,
    pub modules: BTreeMap<String, Module>,
}

#[derive(Debug)]
pub struct Module {
    pub attributes: BTreeMap<String, Value>,
    pub ports: BTreeMap<String, Value>,
    pub cells: BTreeMap<String, Cell>,
    pub netnames: BTreeMap<String, Value>,
}

#[derive(Debug, Deserialize)]
pub struct Cell {
    #[serde(rename = "type")]
    pub kind: String,
}

#[derive(Debug, Deserialize)]
struct RawRtlilJson {
    modules: Option<BTreeMap<String, RawModule>>,
}

#[derive(Debug, Deserialize)]
struct RawModule {
    #[serde(default)]
    attributes: BTreeMap<String, Value>,
    ports: Option<BTreeMap<String, Value>>,
    cells: Option<BTreeMap<String, Cell>>,
    netnames: Option<BTreeMap<String, Value>>,
}

impl Module {
    fn from_raw(name: &str, raw: RawModule) -> Result<Self> {
        let ports = raw
            .ports
            .ok_or_else(|| anyhow!("RTLIL module `{}` is missing `ports`", name))?;
        let cells = raw
            .cells
            .ok_or_else(|| anyhow!("RTLIL module `{}` is missing `cells`", name))?;
        let netnames = raw
            .netnames
            .ok_or_else(|| anyhow!("RTLIL module `{}` is missing `netnames`", name))?;

        Ok(Self {
            attributes: raw.attributes,
            ports,
            cells,
            netnames,
        })
    }

    fn is_top(&self) -> bool {
        const TOP_KEYS: [&str; 2] = ["top", "\\top"];
        self.attributes
            .iter()
            .filter(|(key, _)| TOP_KEYS.contains(&key.as_str()))
            .any(|(_, value)| value_is_truthy(value))
    }
}

pub fn load_rtlil_json(path: impl AsRef<Path>, options: &LoaderOptions) -> Result<RtlilJson> {
    let path = path.as_ref();
    let raw_data = fs::read_to_string(path)
        .with_context(|| format!("failed to read RTLIL JSON from {}", display_path(path)))?;

    let raw: RawRtlilJson = serde_json::from_str(&raw_data)
        .with_context(|| format!("failed to parse RTLIL JSON from {}", display_path(path)))?;

    let modules_map = raw
        .modules
        .ok_or_else(|| anyhow!("RTLIL JSON is missing the `modules` map"))?;
    ensure!(
        !modules_map.is_empty(),
        "RTLIL JSON does not contain any modules"
    );

    let mut modules = BTreeMap::new();
    let mut top_modules = Vec::new();
    let mut memory_cells = Vec::new();

    for (name, raw_module) in modules_map {
        let module = Module::from_raw(&name, raw_module)?;

        if module.is_top() {
            top_modules.push(name.clone());
        }

        for (cell_name, cell) in &module.cells {
            if is_memory_cell(&cell.kind) {
                memory_cells.push(format!("{name}.{cell_name}"));
            }
        }

        modules.insert(name, module);
    }

    let top = match top_modules.len() {
        0 => {
            bail!(
                "RTLIL JSON does not identify a top module. Add the `(* top *)` attribute before exporting."
            )
        }
        1 => top_modules.remove(0),
        _ => bail!(
            "multiple RTLIL modules are marked as top: {}",
            top_modules.join(", ")
        ),
    };

    if !memory_cells.is_empty() && !options.allow_mem_blackbox {
        bail!(
            "RTLIL JSON still contains memory cells that must be lowered: {}",
            memory_cells.join(", ")
        );
    }

    Ok(RtlilJson { top, modules })
}

fn is_memory_cell(cell_type: &str) -> bool {
    matches!(cell_type, "$mem" | "$memrd" | "$memwr")
}

fn value_is_truthy(value: &Value) -> bool {
    match value {
        Value::Bool(true) => true,
        Value::Number(num) => num.as_u64() == Some(1),
        Value::String(s) => s == "1" || s.eq_ignore_ascii_case("true"),
        _ => false,
    }
}

fn display_path(path: &Path) -> String {
    path.display().to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn truthy_attribute_variants() {
        assert!(value_is_truthy(&Value::Bool(true)));
        assert!(value_is_truthy(&Value::String("1".to_string())));
        assert!(value_is_truthy(&Value::String("TRUE".to_string())));
        assert!(value_is_truthy(&serde_json::json!(1)));
        assert!(!value_is_truthy(&Value::Bool(false)));
        assert!(!value_is_truthy(&serde_json::json!(0)));
        assert!(!value_is_truthy(&Value::String("no".to_string())));
    }

    #[test]
    fn memory_cell_detection() {
        assert!(is_memory_cell("$mem"));
        assert!(is_memory_cell("$memrd"));
        assert!(is_memory_cell("$memwr"));
        assert!(!is_memory_cell("$dff"));
    }
}
