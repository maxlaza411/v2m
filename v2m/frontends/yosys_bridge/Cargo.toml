[package]
name = "frontends-yosys-bridge"
version = "0.1.0"
edition = "2021"
license = "MIT"

[dependencies]
anyhow = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
yosys-bridge = { path = "../../crates/yosys-bridge" }
v2m-formats = { path = "../../core/formats" }

[dev-dependencies]
insta = { workspace = true }
rand = { workspace = true }
v2m-evaluator = { path = "../../evaluator" }
