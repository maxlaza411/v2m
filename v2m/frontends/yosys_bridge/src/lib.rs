pub mod cells;
mod context;
mod module;
mod nir_node;

pub use module::{rtlil_to_nir, ModuleBuilder};
pub use nir_node::NirNode;
