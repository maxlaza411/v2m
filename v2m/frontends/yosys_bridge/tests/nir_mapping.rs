use std::collections::BTreeMap;

use anyhow::Result;
use frontends_yosys_bridge::rtlil_to_nir;
use insta::assert_json_snapshot;
use serde_json::{json, Value};
use yosys_bridge::loader::{Cell, Module, RtlilJson};

#[test]
fn alu4_sample_to_nir_snapshot() -> Result<()> {
    let design = alu4_design();
    let nir = rtlil_to_nir(&design)?;
    let value = serde_json::to_value(&nir)?;
    assert_json_snapshot!("alu4_sample", value);
    Ok(())
}

#[test]
fn mux_tree_sample_to_nir_snapshot() -> Result<()> {
    let design = mux_tree_design();
    let nir = rtlil_to_nir(&design)?;
    let value = serde_json::to_value(&nir)?;
    assert_json_snapshot!("mux_tree_sample", value);
    Ok(())
}

fn alu4_design() -> RtlilJson {
    let mut ports = BTreeMap::new();
    ports.insert("a".to_string(), port(&[1, 2, 3, 4], "input"));
    ports.insert("b".to_string(), port(&[5, 6, 7, 8], "input"));
    ports.insert("op".to_string(), port(&[9, 10], "input"));
    ports.insert("y".to_string(), port(&[35, 36, 37, 38], "output"));

    let mut netnames = BTreeMap::new();
    netnames.insert("a".to_string(), net(&[1, 2, 3, 4]));
    netnames.insert("b".to_string(), net(&[5, 6, 7, 8]));
    netnames.insert("op".to_string(), net(&[9, 10]));
    netnames.insert("sum".to_string(), net(&[11, 12, 13, 14]));
    netnames.insert("and_ab".to_string(), net(&[15, 16, 17, 18]));
    netnames.insert("or_ab".to_string(), net(&[19, 20, 21, 22]));
    netnames.insert("xor_ab".to_string(), net(&[23, 24, 25, 26]));
    netnames.insert("lo_sel".to_string(), net(&[27, 28, 29, 30]));
    netnames.insert("hi_sel".to_string(), net(&[31, 32, 33, 34]));
    netnames.insert("y".to_string(), net(&[35, 36, 37, 38]));

    let mut cells = BTreeMap::new();
    cells.insert(
        "add_sum".to_string(),
        add_cell(
            &[
                ("A", &[1, 2, 3, 4]),
                ("B", &[5, 6, 7, 8]),
                ("Y", &[11, 12, 13, 14]),
            ],
            4,
            4,
            4,
        ),
    );
    cells.insert(
        "and_node".to_string(),
        bitwise_cell(
            "$and",
            &[
                ("A", &[1, 2, 3, 4]),
                ("B", &[5, 6, 7, 8]),
                ("Y", &[15, 16, 17, 18]),
            ],
            4,
        ),
    );
    cells.insert(
        "or_node".to_string(),
        bitwise_cell(
            "$or",
            &[
                ("A", &[1, 2, 3, 4]),
                ("B", &[5, 6, 7, 8]),
                ("Y", &[19, 20, 21, 22]),
            ],
            4,
        ),
    );
    cells.insert(
        "xor_node".to_string(),
        bitwise_cell(
            "$xor",
            &[
                ("A", &[1, 2, 3, 4]),
                ("B", &[5, 6, 7, 8]),
                ("Y", &[23, 24, 25, 26]),
            ],
            4,
        ),
    );
    cells.insert(
        "mux_lo".to_string(),
        mux_cell(
            &[
                ("A", &[11, 12, 13, 14]),
                ("B", &[15, 16, 17, 18]),
                ("S", &[9]),
                ("Y", &[27, 28, 29, 30]),
            ],
            4,
            1,
        ),
    );
    cells.insert(
        "mux_hi".to_string(),
        mux_cell(
            &[
                ("A", &[19, 20, 21, 22]),
                ("B", &[23, 24, 25, 26]),
                ("S", &[9]),
                ("Y", &[31, 32, 33, 34]),
            ],
            4,
            1,
        ),
    );
    cells.insert(
        "mux_final".to_string(),
        mux_cell(
            &[
                ("A", &[27, 28, 29, 30]),
                ("B", &[31, 32, 33, 34]),
                ("S", &[10]),
                ("Y", &[35, 36, 37, 38]),
            ],
            4,
            1,
        ),
    );

    let module = Module {
        attributes: BTreeMap::new(),
        ports,
        cells,
        netnames,
    };

    RtlilJson {
        top: "alu4".to_string(),
        modules: BTreeMap::from([(module_name("alu4"), module)]),
    }
}

fn mux_tree_design() -> RtlilJson {
    let mut ports = BTreeMap::new();
    ports.insert("in0".to_string(), port(&[100, 101, 102, 103], "input"));
    ports.insert("in1".to_string(), port(&[104, 105, 106, 107], "input"));
    ports.insert("in2".to_string(), port(&[108, 109, 110, 111], "input"));
    ports.insert("in3".to_string(), port(&[112, 113, 114, 115], "input"));
    ports.insert("sel".to_string(), port(&[116, 117], "input"));
    ports.insert("y".to_string(), port(&[140, 141, 142, 143], "output"));

    let mut netnames = BTreeMap::new();
    netnames.insert("in0".to_string(), net(&[100, 101, 102, 103]));
    netnames.insert("in1".to_string(), net(&[104, 105, 106, 107]));
    netnames.insert("in2".to_string(), net(&[108, 109, 110, 111]));
    netnames.insert("in3".to_string(), net(&[112, 113, 114, 115]));
    netnames.insert("sel".to_string(), net(&[116, 117]));
    netnames.insert("not_in0".to_string(), net(&[118, 119, 120, 121]));
    netnames.insert("xnor_in1_in2".to_string(), net(&[122, 123, 124, 125]));
    netnames.insert("diff".to_string(), net(&[126, 127, 128, 129]));
    netnames.insert("mux_lo".to_string(), net(&[130, 131, 132, 133]));
    netnames.insert("mux_hi".to_string(), net(&[134, 135, 136, 137]));
    netnames.insert("y".to_string(), net(&[140, 141, 142, 143]));

    let mut cells = BTreeMap::new();
    cells.insert(
        "not_in0".to_string(),
        not_cell(
            &[("A", &[100, 101, 102, 103]), ("Y", &[118, 119, 120, 121])],
            4,
        ),
    );
    cells.insert(
        "xnor_inputs".to_string(),
        bitwise_cell(
            "$xnor",
            &[
                ("A", &[104, 105, 106, 107]),
                ("B", &[108, 109, 110, 111]),
                ("Y", &[122, 123, 124, 125]),
            ],
            4,
        ),
    );
    cells.insert(
        "sub_inputs".to_string(),
        sub_cell(
            &[
                ("A", &[112, 113, 114, 115]),
                ("B", &[108, 109, 110, 111]),
                ("Y", &[126, 127, 128, 129]),
            ],
            4,
            4,
            4,
        ),
    );
    cells.insert(
        "mux_level0".to_string(),
        mux_cell(
            &[
                ("A", &[118, 119, 120, 121]),
                ("B", &[122, 123, 124, 125]),
                ("S", &[116]),
                ("Y", &[130, 131, 132, 133]),
            ],
            4,
            1,
        ),
    );
    cells.insert(
        "mux_level1".to_string(),
        mux_cell(
            &[
                ("A", &[126, 127, 128, 129]),
                ("B", &[112, 113, 114, 115]),
                ("S", &[116]),
                ("Y", &[134, 135, 136, 137]),
            ],
            4,
            1,
        ),
    );
    cells.insert(
        "mux_final".to_string(),
        mux_cell(
            &[
                ("A", &[130, 131, 132, 133]),
                ("B", &[134, 135, 136, 137]),
                ("S", &[117]),
                ("Y", &[140, 141, 142, 143]),
            ],
            4,
            1,
        ),
    );

    let module = Module {
        attributes: BTreeMap::new(),
        ports,
        cells,
        netnames,
    };

    RtlilJson {
        top: "mux_tree".to_string(),
        modules: BTreeMap::from([(module_name("mux_tree"), module)]),
    }
}

fn add_cell(connections: &[(&str, &[i64])], a_width: u32, b_width: u32, y_width: u32) -> Cell {
    let mut parameters = BTreeMap::new();
    parameters.insert("A_WIDTH".to_string(), json!(a_width));
    parameters.insert("B_WIDTH".to_string(), json!(b_width));
    parameters.insert("Y_WIDTH".to_string(), json!(y_width));
    parameters.insert("A_SIGNED".to_string(), json!(false));
    parameters.insert("B_SIGNED".to_string(), json!(false));
    parameters.insert("Y_SIGNED".to_string(), json!(false));
    make_cell("$add", parameters, connections)
}

fn sub_cell(connections: &[(&str, &[i64])], a_width: u32, b_width: u32, y_width: u32) -> Cell {
    let mut parameters = BTreeMap::new();
    parameters.insert("A_WIDTH".to_string(), json!(a_width));
    parameters.insert("B_WIDTH".to_string(), json!(b_width));
    parameters.insert("Y_WIDTH".to_string(), json!(y_width));
    parameters.insert("A_SIGNED".to_string(), json!(false));
    parameters.insert("B_SIGNED".to_string(), json!(false));
    parameters.insert("Y_SIGNED".to_string(), json!(false));
    make_cell("$sub", parameters, connections)
}

fn bitwise_cell(kind: &str, connections: &[(&str, &[i64])], width: u32) -> Cell {
    let mut parameters = BTreeMap::new();
    parameters.insert("A_WIDTH".to_string(), json!(width));
    parameters.insert("B_WIDTH".to_string(), json!(width));
    parameters.insert("Y_WIDTH".to_string(), json!(width));
    make_cell(kind, parameters, connections)
}

fn not_cell(connections: &[(&str, &[i64])], width: u32) -> Cell {
    let mut parameters = BTreeMap::new();
    parameters.insert("A_WIDTH".to_string(), json!(width));
    parameters.insert("Y_WIDTH".to_string(), json!(width));
    make_cell("$not", parameters, connections)
}

fn mux_cell(connections: &[(&str, &[i64])], width: u32, select_width: u32) -> Cell {
    let mut parameters = BTreeMap::new();
    parameters.insert("WIDTH".to_string(), json!(width));
    parameters.insert("S_WIDTH".to_string(), json!(select_width));
    make_cell("$mux", parameters, connections)
}

fn make_cell(
    kind: &str,
    parameters: BTreeMap<String, Value>,
    connections: &[(&str, &[i64])],
) -> Cell {
    let mut connections_map = BTreeMap::new();
    for (pin, bits) in connections {
        connections_map.insert((*pin).to_string(), json!(bits));
    }

    Cell {
        kind: kind.to_string(),
        parameters,
        attributes: BTreeMap::new(),
        connections: connections_map,
    }
}

fn port(bits: &[i64], direction: &str) -> Value {
    json!({ "direction": direction, "bits": bits })
}

fn net(bits: &[i64]) -> Value {
    json!({ "bits": bits })
}

fn module_name(name: &str) -> String {
    name.to_string()
}
