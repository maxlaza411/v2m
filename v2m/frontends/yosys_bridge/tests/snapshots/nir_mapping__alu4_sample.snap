---
source: frontends/yosys_bridge/tests/nir_mapping.rs
assertion_line: 14
expression: value
---
{
  "design": "alu4",
  "modules": {
    "alu4": {
      "nets": {
        "a": {
          "bits": 4
        },
        "and_ab": {
          "bits": 4
        },
        "b": {
          "bits": 4
        },
        "hi_sel": {
          "bits": 4
        },
        "lo_sel": {
          "bits": 4
        },
        "op": {
          "bits": 2
        },
        "or_ab": {
          "bits": 4
        },
        "sum": {
          "bits": 4
        },
        "xor_ab": {
          "bits": 4
        },
        "y": {
          "bits": 4
        }
      },
      "nodes": {
        "add_sum": {
          "attrs": {
            "A_SIGNED": false,
            "A_WIDTH": 4,
            "B_SIGNED": false,
            "B_WIDTH": 4,
            "Y_SIGNED": false,
            "Y_WIDTH": 4
          },
          "op": "ADD",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 3,
              "net": "a"
            },
            "B": {
              "lsb": 0,
              "msb": 3,
              "net": "b"
            },
            "Y": {
              "lsb": 0,
              "msb": 3,
              "net": "sum"
            }
          },
          "uid": "add_sum",
          "width": 4
        },
        "and_node": {
          "op": "AND",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 3,
              "net": "a"
            },
            "B": {
              "lsb": 0,
              "msb": 3,
              "net": "b"
            },
            "Y": {
              "lsb": 0,
              "msb": 3,
              "net": "and_ab"
            }
          },
          "uid": "and_node",
          "width": 4
        },
        "mux_final": {
          "op": "MUX",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 3,
              "net": "lo_sel"
            },
            "B": {
              "lsb": 0,
              "msb": 3,
              "net": "hi_sel"
            },
            "S": {
              "lsb": 1,
              "msb": 1,
              "net": "op"
            },
            "Y": {
              "lsb": 0,
              "msb": 3,
              "net": "y"
            }
          },
          "uid": "mux_final",
          "width": 4
        },
        "mux_hi": {
          "op": "MUX",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 3,
              "net": "or_ab"
            },
            "B": {
              "lsb": 0,
              "msb": 3,
              "net": "xor_ab"
            },
            "S": {
              "lsb": 0,
              "msb": 0,
              "net": "op"
            },
            "Y": {
              "lsb": 0,
              "msb": 3,
              "net": "hi_sel"
            }
          },
          "uid": "mux_hi",
          "width": 4
        },
        "mux_lo": {
          "op": "MUX",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 3,
              "net": "sum"
            },
            "B": {
              "lsb": 0,
              "msb": 3,
              "net": "and_ab"
            },
            "S": {
              "lsb": 0,
              "msb": 0,
              "net": "op"
            },
            "Y": {
              "lsb": 0,
              "msb": 3,
              "net": "lo_sel"
            }
          },
          "uid": "mux_lo",
          "width": 4
        },
        "or_node": {
          "op": "OR",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 3,
              "net": "a"
            },
            "B": {
              "lsb": 0,
              "msb": 3,
              "net": "b"
            },
            "Y": {
              "lsb": 0,
              "msb": 3,
              "net": "or_ab"
            }
          },
          "uid": "or_node",
          "width": 4
        },
        "xor_node": {
          "op": "XOR",
          "pin_map": {
            "A": {
              "lsb": 0,
              "msb": 3,
              "net": "a"
            },
            "B": {
              "lsb": 0,
              "msb": 3,
              "net": "b"
            },
            "Y": {
              "lsb": 0,
              "msb": 3,
              "net": "xor_ab"
            }
          },
          "uid": "xor_node",
          "width": 4
        }
      },
      "ports": {
        "a": {
          "bits": 4,
          "dir": "input"
        },
        "b": {
          "bits": 4,
          "dir": "input"
        },
        "op": {
          "bits": 2,
          "dir": "input"
        },
        "y": {
          "bits": 4,
          "dir": "output"
        }
      }
    }
  },
  "top": "alu4",
  "v": "nir-1.1"
}
